### PR TITLE
RFC: Bundle type

### PR DESCRIPTION
**Pull request by thibault-martinez**
[Rendered](https://github.com/thibault-martinez/bee-rfcs/blob/bee-bundle/text/0018-bee-bundle/0018-bee-bundle.md)

### CODE REVIEW

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
++ Feature name: `bee-bundle`
++ Start date: 2019-10-02
++ RFC PR: [iotaledger/bee-rfcs#18](https://github.com/iotaledger/bee-rfcs/pull/18)
++ Bee issue: [iotaledger/bee#63](https://github.com/iotaledger/bee/issues/63)
+
+# Summary
+
+The smallest communication unit in the IOTA protocol is the transaction. Everything, including payment settlements and/or plain data, is propagated through the IOTA network in transactions.
+
+A transaction is `2673` trytes and the part available to the user is `2187` trytes. This part holds a signature in case of a payment settlement and plain data otherwise. Since it has a limited size, a user often needs more than one transaction to fulfil his operation, for example signatures with security level `2` or `3` don't fit in a single transaction and user-provided data may exceed the allowance so they need to be fragmented across multiple transactions. Moreover, a transaction may not make sense on its own, for example an input/output transaction alone would change the total amount of the ledger so it has to be paired with another complementary input/output.
```
Comment (tsvisabo) [inline text/0018-bee-bundle/0018-bee-bundle.md]: "his" => "his/her"
"a transaction may not make sense on its own" => "a value transaction will not make sense on its own because .."

Comment (tsvisabo) [inline text/0018-bee-bundle/0018-bee-bundle.md]: "..complementary input/output"  + "that together will balance the balance total value to zero"

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+| Name          | Size      |
+| ------------- | --------- |
+| address       | 243 trits |
+| value         | 81 trits  |
+| obsolete_tag  | 81 trits  |
+| timestamp     | 27 trits  |
+| current_index | 27 trits  |
+| last_index    | 27 trits  |
+
+The bundle hash is generated with a sponge by iterating through the bundle, from `0` to `last_index`, absorbing the bundle essence of each transaction and eventually squeezing the bundle hash from the sponge.
```
Comment (tsvisabo) [inline text/0018-bee-bundle/0018-bee-bundle.md]: sponge function

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+Validating a bundle means checking the syntactic and semantic integrity of a bundle as a whole and of its constituent transactions. As bundles are atomic transfers, either all or none of the transactions will be accepted by the network. After validation, transactions of a bundle are candidates to be included to the ledger.
+
+For a bundle to be considered valid, the following assertions must be true:
+
+- bundle has announced size;
+- transactions share the same bundle hash;
+- transactions absolute value doesn't exceed total IOTA supply;
+- bundle absolute sum never exceeds total IOTA supply;
+- transactions appear in the announced order;
```
Comment (tsvisabo) [inline text/0018-bee-bundle/0018-bee-bundle.md]: what does it mean `announced`? index?
what does it mean `appear`?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 1 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Thanks, I'll go with they/their which is usually the recommended gender neutral form.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 2 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Technically, I'm speaking of a sponge object as seen in the pseudocode, so I think it's fine.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 3 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Rephrased

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
++ Bee issue: [iotaledger/bee#63](https://github.com/iotaledger/bee/issues/63)
+
+# Summary
+
+This RFC is based on [`transaction-module`]().
+<!-- TODO FIx link to RFC#3 -->
+
+The smallest communication unit in the IOTA protocol is the transaction. Everything, including payment settlements and/or plain data, is propagated through the IOTA network in transactions.
+
+A transaction is `2673` trytes and the part available to the user is `2187` trytes. This part holds a signature in case of a payment settlement and plain data otherwise. Since it has a limited size, a user often needs more than one transaction to fulfil their operation, for example signatures with security level `2` or `3` don't fit in a single transaction and user-provided data may exceed the allowance so they need to be fragmented across multiple transactions. Moreover, a value transaction doesn't make sense on its own because it would change the total amount of the ledger so it has to be paired with other complementary transactions that together will balance the total value to zero.
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Let's express this in `trits`. Further down, the RFC talks about `trits` instead of `trytes`. I think `trits` is the more natural/less confusing unit to talk about. `trit` already puts some mental overhead on the user, because it's very unusual, and `tryte` being defined as a multiple of `trit` even more so.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+For these reasons, transactions have to be processed as a whole, in groups called bundles. A bundle is an atomic operation in the sense that either all or none of its transactions are accepted by the network. Even single transactions are propagated through the network within a bundle making it the only confirmable communication unit of the IOTA protocol.
+
+This RFC proposes ways to create and manipulate a bundle and describe the associated algorithms.
+
+Useful links:
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:25]: Maybe we should also put a link to trits here, or at least quickly define them?

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+### Bundle
+
+As bundles are final, they shouldn't be modifiable outside of the scope of the bundle module.
+
+A bundle can simply be represented as an array of transactions:
+
+```rust
+struct Bundle {
+    transactions: Vec<Transaction>
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Is there a natural order to `transactions`? Should they be included one transaction at a time? Is there a chance that transactions arrive out of order? Can we use a `HashMap` instead of `Transaction` here?

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+}
+```
+
+And its implementation should only allow to retrieve transactions:
+
+```rust
+impl Bundle {    
+    pub fn transactions(&self) -> &Vec<Transaction> {
+        &self.transactions
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:77]: Should we hide this as an implementation detail and instead provide a `struct Transactions(Vec<Transaction>)` newtype?

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+}
+```
+
+### BundleBuilder
+
+The `BundleBuilder` offers a simple and convenient way to build bundles:
+
+```rust
+pub struct BundleBuilder {
+    transactions: Vec<TransactionBuilder>
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:88]: Is there a semantic difference between an empty vector and a vector that hasn't been set? If yes, I would set this to `Option<Vec<TransactionBuilder>>` (or to `Option<TransactionBuilders>`, if we want to hide the details).

Please also note my proposal for constructing a `Transaction` via a `UnprovenTransaction` at https://github.com/rufsam/bee-rfcs/pull/1

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+    }
+}
+```
+
+*We do not list parameters and/or return values as they are implementation details.*
+
+## Algorithms
+
+In this section, we describe the algorithms needed to build a `Bundle`. The lifecycle of a `BundleBuilder` depends on if it's being used in client side or server side:
+- client side: `finalise` -> [`sign` ->] -> [`pow` ->] -> `validate` -> `build`
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Corrected"
`finalise -> [sign ->] [pow ->] validate -> build`

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+- server side: `addTransaction`/`addTransactionBuilder` -> `validate` -> `build`
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Rust convention is `small_snake_case`, `add_transaction` and `add_transaction_builder`.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+*`sign` is optional because data transactions don't have to be signed. `pow` is optional because one can use remote pow instead.*
+
+### Hash
+
+*Client side and server side operation.*
+
+A bundle hash ties different transactions together. By having this common hash in their `bundle` field, it makes it clear that these transactions should be processed as a whole.
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: If I read this correctly, this is a hard requirement on the design of `Transaction`. If that's the case, we might need to combine the RFCs after all.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+The hash of a bundle is derived from the bundle essence of each of its transactions. The bundle essence of a transaction is composed of the following fields.
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: We should make it explicit that `bundle essence` refers to a subset of fields comprising the transaction.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+
+| Name          | Size      |
+| ------------- | --------- |
+| address       | 243 trits |
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:144]: For consistency, we should decide whether we want to use `trits` or `trytes` here (see also the review comment of the summary section).

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+Pseudocode:
+
+```
+hash(bundle)
+| sponge = Sponge(HASH_FUNCTION)
+|
+| for transaction in bundle
+| | sponge.absorb(transaction.essence())
+|
+| return sponge.squeeze()
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:163]: I'd prefer to not have pipes in here, as the whitespace is sufficient to distinguish blocks.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+|
+| while hash.normalise().find('M')
+| | bundle.at(0).obsolete_tag++
+| | hash = bundleHash(bundle)
+|
+| for transaction in bundle
+| | transaction.setBundleHash(hash)
+```
+
+*Security requirement: due to the implementation of the signature process, the normalised bundle hash can't contain a `M` or `13` because it could expose a significant part of the private key, weakening the signature. The bundle hash is then repetitively generated with a slight modification until its normalisation doesn't contain a `M`.*
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: This is a sufficiently weird caveat that we should definitely further explain it or point to the revelant documents explaining it.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+
+```rust
+pub struct BundleBuilder {
+    transactions: Vec<TransactionBuilder>
+}
+```
+
+```rust
+impl BundleBuilder {    
+    pub fn hash(&self) {
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: In cases where the verb isn't clear, I would opt to prepend a method with `calculate_`, making this `calculate_hash`.

Also for expensive operations it signifies that some work will be done.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+
+    pub fn finalise(&self) {
+      unimplemented!()
+    }
+
+    pub fn sign(&self) {
+      unimplemented!()
+    }
+
+    pub fn pow(&self) {
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md]: As for `hash`, I would call this `calculate_proof_of_work`. We might consider passing a function `F: Fn(&[u8]) -> Nonce` here to allow for different PoW algorithms.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+
+*Client side operation.*
+
+Finalising a bundle means computing the bundle hash, verifying that it matches the security requirement and setting it to all the transactions of the bundle. After finalisation, transactions of a bundle are ready to be safely attached to the tangle.
+
+Pseudocode:
+
+```
+finalise(bundle)
+| hash = bundleHash(bundle)
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:178]: Same as above, for these short blocks I think the extra `|` produce too much noise.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+- order of transactions in the bundle is the same as announced by `current_index` and `last_index`;
+- value transactions have an address ending in `0` i.e. has been generated with Kerl;
+- bundle inputs and outputs are balanced i.e. the bundle sum equals `0`;
+- announced bundle hash matches the computed bundle hash;
+- for spending transactions, the signature is valid;
+
+Pseudocode:
+
+```
+validate(bundle):
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:270]: Because a `Bundle` should be a valid object, the validation step should be part of the `build` method.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+
+# Rationale and alternatives
+
+- A `Bundle` is a fundamental component of the IOTA protocol and must be implemented;
+- There is no more intuitive and simple way to implement a `Bundle` than the one proposed;
+- Since bundles are final, `BundleBuilder` is mandatory;
+
+# Unresolved questions
+
+- Should this RFC expands a bit more on the M-Bug ? Or give a link ?
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:317]: Yes.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+- Should `Bundle` provide a `.transactions` or a `.at` method ?
```
Comment (SuperFluffy) [inline text/0018-bee-bundle/0018-bee-bundle.md:318]: Is there a natural order to `Bundle`? Can `Transactions` arrive out of order?

The `.at` method is basically indexing into the `Bundle`, getting a specific transaction? I think this makes sense, if there is a natural order.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 4 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Will do it for the sake of consistency. But when talking about user-facing sizes we usually prefer trytes over trits because of the nice readable mapping. And trits are usually used in crypto contexts.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 5 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:25]: Sure

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 6 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: There is a natural order. Each transaction has a `current_index` and a `last_index`. So transactions in a bundle should be ordered from `0` to `last_index`.

> Is there a chance that transactions arrive out of order

Yes definitely, but we don't really care because:
- client-side: we add transaction one at a time
- server-side: we process bundle only when we know all the transactions are present

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 7 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:77]: Good idea

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 8 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:88]: I don't think there is a difference, in both case it'll fail to build a bundle

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 9 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Whoops, thanks

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 10 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Thanks

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 14 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:163]: Algorithms in papers usually use bars but alright, whitespaces

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 15 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Yeah I know :/
But at the same time this should and will be expanded in the actual Signing Scheme RFC
The perfect way would be later to link to the RFC

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 16 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Sure

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 19 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:270]: `build` will definitely call `validate`. But I assumed `validate` is a function sufficiently interesting be able to be called on its own. I can imagine situation when we want to validate but doesn't care of the resulting bundle.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 21 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:318]: Yes there is, `.at()` it is then.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 4 above)
Comment (samuel-rufi) [reply text/0018-bee-bundle/0018-bee-bundle.md]: > value transaction doesn't make sense on its own

maybe call it` payment settlement` instead `value transaction`. We should distinguish `transaction`, the atomic unit, from a settlement which requires multiple transactions. The term "transaction" should not be used as synonym for `bundle`, else it confuses only.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+hash(bundle)
+| sponge = Sponge(HASH_FUNCTION)
+|
+| for transaction in bundle
+| | sponge.absorb(transaction.essence())
+|
+| return sponge.squeeze()
+```
+
+*In the current mainnet, the hash function of the sponge used to generate bundle hashes is Kerl.*
```
Comment (samuel-rufi) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Not sure if we should write "in the current mainnet". I mean this bundle proposal is directly for the mainnet. We shouldn't need to refer it.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 12 above)
Comment (samuel-rufi) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Agree, and why it's not possible to use all transaction fields. Moreover a reader might ask why `bundle essence` and not simply `bundle`.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 6 above)
Comment (wusyong) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Yes, we might receive the transaction out of order. I think stay in vector makes sense. Otherwise, we shouldn't just use `HashMap` and should build our own linked list furthermore.

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+
+*We do not list parameters and/or return values as they are implementation details.*
+
+## Algorithms
+
+In this section, we describe the algorithms needed to build a `Bundle`. The lifecycle of a `BundleBuilder` depends on if it's being used in client side or server side:
+- client side: `finalise` -> [`sign` ->] -> [`pow` ->] -> `validate` -> `build`
+- server side: `addTransaction`/`addTransactionBuilder` -> `validate` -> `build`
+
+*`sign` is optional because data transactions don't have to be signed. `pow` is optional because one can use remote pow instead.*
```
Comment (wusyong) [inline text/0018-bee-bundle/0018-bee-bundle.md]: As we introduce intermediate type, I think pow validation will be mandatory before built? Besides, even it's remote pow, client still needs to wait for returned nonce no?

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 11 above)
Comment (wusyong) [reply text/0018-bee-bundle/0018-bee-bundle.md]: There are also other ways to form a bundle like `Entry`. While I'm not sure we are also going this way, I would suggest these should be separate RFC.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 4 above)
Comment (wusyong) [reply text/0018-bee-bundle/0018-bee-bundle.md]: I would recommend stay in trytes and maybe just mention it's 3 trits 1 trytes. Many documentations and other libraries across IOTA ecosystem are still in form of trytes when it comes to build transaction and bundles.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 17 above)
Comment (wusyong) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Might also want some trait bound or trait object. But I guess this depends on design of POW module. Let's just stay this for now. Further discussion should bring to #19.

Comment (wusyong) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Btw, I think we should pass `mwm` too.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 4 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: @rufsam But here `value transaction` is used on purpose to indeed make the point that we need bundles.

Comment (samuel-rufi) [reply text/0018-bee-bundle/0018-bee-bundle.md]: @thibault-martinez okey understand what you mean. however, in general, we shouldn't use the term `value transaction` for transfering value, and maybe make this clear in the RFC.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 7 above)
Comment (zesterer) [reply text/0018-bee-bundle/0018-bee-bundle.md:77]: Alternatively, you could return either `&[Transaction]` or `impl FixedSizeIterator<Item=&Transaction>`

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+
+# Summary
+
+This RFC is based on [`transaction-module`]().
+<!-- TODO FIx link to RFC#3 -->
+
+The smallest communication unit in the IOTA protocol is the transaction. Everything, including payment settlements
+and/or plain data, is propagated through the IOTA network in transactions.
+
+A transaction is `2673` trytes and the part available to the user is `2187` trytes. This part holds a signature in case
```
Comment (Alex6323) [inline text/0018-bee-bundle/0018-bee-bundle.md]: * maybe we should also mention the byte equivalent? ~1605 bytes (in the most compact encoding)
* "the part available to the user" -> should we just use the common term "payload" to describe that?

Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Janis wants trits, you want bytes, Yu-Wei wants trytes... :trollface: 
Thanks, I like payload.

Comment (Alex6323) [reply text/0018-bee-bundle/0018-bee-bundle.md]: regarding the byte equivalent: I had my IOTA-newcomer glasses on :smile:

File: text/0018-bee-bundle/0018-bee-bundle.md [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+of a payment settlement and plain data otherwise. Since it has a limited size, a user often needs more than one
+transaction to fulfil their operation, for example signatures with security level `2` or `3` don't fit in a single
+transaction and user-provided data may exceed the allowance so they need to be fragmented across multiple transactions.
```
Comment (Alex6323) [inline text/0018-bee-bundle/0018-bee-bundle.md:14]: This is gonna be a bit controversial, but I feel this is the only point in time where we can change a fundamental term:

Isn't essentially the `bundle` what should be called a `transaction` according to the database definition of a transaction where a transaction is an *atomic* operation. But in IOTA's case it is the bundle that is atomic, not the transaction, right?
Atomicity: A transaction must be fully complete, saved (committed) or completely undone (rolled back). A sale in a retail store database illustrates a scenario which explains atomicity, e.g., the sale consists of an inventory reduction and a record of incoming cash. Both either happen together or do not happen - it's all or nothing.

Likewise either all the transactions in a bundle get confirmed, or none of them.

The reasoning for bringing this up is again looking on the proposal from the outside as someone who might not be familiar with IOTA terms, and making its terminology most accessible to newcomers. 

It doesn't sound as sexy but I would propose the terms `Transaction` (former bundle) and `Transaction Fragment` (former transaction), while the `Signature Or Message Fragment` is renamed to `Payload`.

If I have made an essential mistake here, then please correct me. Wouldn't be the first time :trollface:

Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:14]: You are 100% right and I've had this thought too. We could have `Transaction` and `Statement` to match the database metaphor. Unfortunately you're  also right saying it is controversial. Not entirely sure we can change terms at this point ?

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 24 above)
Comment (SuperFluffy) [reply text/0018-bee-bundle/0018-bee-bundle.md]: If we go for something exotic, then it should be `trits`. `trytes` being defined in terms of `trits` is an extra layer on top of `trits`, so a user would need to keep both `trits` and `trytes` in mind.

I very much like bytes as well, but since then we also need to talk about compact encodings we should probably not do that.

Comment (Thoralf-M) [inline text/0018-bee-bundle/0018-bee-bundle.md]: Shouldn't the tag also be available to the user?

Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: Yes you are right. I'll rephrase, I was referring only to sig/msg :)

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 14 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md:163]: I tried. Honestly with nested loops it's pretty hard to follow, pipes make it really easy. I'd rather keep them.

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 22 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: You're right

File: text/0018-bee-bundle/0018-bee-bundle.md (hunk 17 above)
Comment (thibault-martinez) [reply text/0018-bee-bundle/0018-bee-bundle.md]: I renamed. But I omitted parameters on purpose, see comment just after the prototypes, because this is implementation detail.

**Comment by SuperFluffy** [comment]
Incorporated into #20.

While discussing the design of the `Transaction` type, see #3, we realized that a new transaction is not supposed to be create directly. Instead, when constructing a new message, the Bundle is the fundamental object one wants to crate, which in turn splits up its message into several Transactions. A Transaction thus becomes a very simple type. More complicated logic should thus be handled by the Bundle.

This makes it necessary to expand the `Bundle` RFC beyond what is proposed here.
